uberenv.py

```diff
@@ -269,6 +290,115 @@ def detect_platform(self):
         return res
 
 
+class VcpkgEnv(UberEnv):
+    """ Helper to clone vcpkg and install libraries on Windows """
+
+    def __init__(self, opts, extra_opts):
+        UberEnv.__init__(self,opts,extra_opts)
+
+        # setup architecture triplet
+        self.triplet = opts["triplet"]
+        if self.triplet is None:
+           self.triplet = os.getenv("VCPKG_DEFAULT_TRIPLET", "x86-windows")
+
+    def setup_paths_and_dirs(self):
+        # get the current working path, and the glob used to identify the
+        # package files we want to hot-copy to vcpkg
+
+        UberEnv.setup_paths_and_dirs(self)
+
+        self.ports = pjoin(self.uberenv_path, "vcpkg_ports","*")
+
+        # setup path for vcpkg repo
+        self.dest_vcpkg = pjoin(self.dest_dir,"vcpkg")
+
+        if os.path.isdir(self.dest_vcpkg):
+            print("[info: destination '{0}' already exists]".format(self.dest_vcpkg))
+
+    def clone_repo(self):
+        if not os.path.isdir(self.dest_vcpkg):
+            # compose clone command for the dest path, vcpkg url and branch
+            vcpkg_branch = self.project_opts.get("vcpkg_branch", "master")
+            vcpkg_url = self.project_opts.get("vcpkg_url", "https://github.com/microsoft/vcpkg")
+
+            print("[info: cloning vcpkg '{0}' branch from {1} into {2}]"
+                .format(vcpkg_branch,vcpkg_url, self.dest_vcpkg))
+
+            os.chdir(self.dest_dir)
+
+            clone_opts = ("-c http.sslVerify=false " 
+                          if self.opts["ignore_ssl_errors"] else "")
+
+            clone_cmd =  "git {0} clone -b {1} {2}".format(clone_opts, vcpkg_branch,vcpkg_url)
+            sexe(clone_cmd, echo=True)
+
+            # optionally, check out a specific commit
+            if "vcpkg_commit" in self.project_opts:
+                sha1 = self.project_opts["vcpkg_commit"]
+                print("[info: using vcpkg commit {0}]".format(sha1))
+                os.chdir(self.dest_vcpkg)
+                sexe("git checkout {0}".format(sha1),echo=True)
+                
+        if self.opts["repo_pull"]:
+            # do a pull to make sure we have the latest
+            os.chdir(self.dest_vcpkg)
+            sexe("git stash", echo=True)
```

Comment:
I know this is trying to be safe but are we expecting people to have changes? Shouldnt we error out in this case?

Reply:
Since vcpkg is patched by Uberenv (without commit) I guess it makes sense?

Reply:
yes i keep forgetting that... i really wish we could get the spack guys to allow us a workflow where we don't have to patch spack..  ignore my comment.

Reply:
if folks have changes, they should be in their copy of the packages or config files (which will get copied over after this cleans up). Doing git stash was the only sane way to clean things up and make it so stuff doesn't turned into a tangled mess run-to-run.

Reply:
(or `git reset --hard HEAD` if you don’t want to keep the changes in the stash list)

Reply:
I believe we need `git stash` b/c the `spack_commit` option can leave is in a detached head state, confusing `reset HEAD`

Reply:
That’s indeed likely to happen. If it were me, I would always create a branch (e.g. with the `date`) in order to avoid any detached head magic.

Reply:
But you could end up with many branches at the same spot... Or create a branch with a unique uberenv-specific name.

Reply:
OK, OK, let’s keep things simple !?